### PR TITLE
[NPUW] Fix eagle3 with chunk prefill

### PR DESCRIPTION
**Pull request by GuoliangShiIntel**
### Details:
**Background:**
1. The Eagle3 Target/Draft model now outputs `last_hidden_status` in addition to `logits`
2. While `logits` only needs the last token (via slice output), `last_hidden_status` requires all token tensors.
3. For chunk prefill, we must accumulate `last_hidden_status` across chunks, unlike `logits` which only needs the final chunk.

**Changes in this PR:**
Added logic to accumulate and concatenate `last_hidden_status` outputs across chunks during chunk prefill in the Eagle3 pipeline.

### Tickets:
[ - *CVS-180647*](https://jira.devtools.intel.com/browse/CVS-180647)

### CODE REVIEW

File: src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.cpp [hunk 1]
```diff
@@ -159,6 +159,68 @@ void Eagle3Extension::update_last_hidden_state(
...
+        m_last_hidden_state = ov::get_tensor_impl(
+            ov::Tensor(chunk_output->get_element_type(), ov::Shape{batch_size, total_seq_len, hidden_size}));
+        m_chunked_seq_offset = 0;
+
+        LOG_VERB("Eagle3: Pre-allocated last_hidden_state tensor with shape=[" << batch_size << "," << total_seq_len
+                                                                               << "," << hidden_size << "]");
+    }
+
+    const auto& target_shape = m_last_hidden_state->get_shape();
+    const uint32_t target_total_len = static_cast<uint32_t>(target_shape[1]);
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.cpp:197]: Should we do `OPENVINO_ASSERT(target_total_len  == total_seq_len)` or it is completely unnecessary?

File: src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.cpp [hunk 2]
```diff
@@ -159,6 +159,68 @@ void Eagle3Extension::update_last_hidden_state(
...
+
+    OPENVINO_ASSERT(m_chunked_seq_offset + chunk_token_count <= target_total_len,
+                    "Chunked sequence offset exceeds pre-allocated size");
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.cpp]: `"Can't write chunk by stored chunked sequence offset and requested number of tokens, as it will exceed pre-allocated size"`

File: src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.cpp [hunk 3]
```diff
@@ -159,6 +159,68 @@ void Eagle3Extension::update_last_hidden_state(
...
+    const uint32_t chunk_start_offset = chunk_seq_len - chunk_token_count;
+    const size_t hidden_elem_size = chunk_output->get_element_type().size();
+    const size_t row_bytes = hidden_size * hidden_elem_size;
+
+    const uint8_t* chunk_ptr = reinterpret_cast<const uint8_t*>(chunk_output->data());
+    chunk_ptr += chunk_start_offset * row_bytes;  // Skip padding, point to valid tokens
+
+    // Copy chunk data directly to the correct position in pre-allocated tensor
+    uint8_t* dst_ptr = reinterpret_cast<uint8_t*>(m_last_hidden_state->data());
+    dst_ptr += m_chunked_seq_offset * row_bytes;  // Move to the current write position
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.cpp]: Could we please use `ov::npuw::util::make_tensor_slice` and `tensor->copy_to(another_tensor)` here? Some examples can be found in `LLMInferRequest`: https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp

File: src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.cpp (hunk 1 above)
Comment (GuoliangShiIntel) [reply src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.cpp:197]: Good suggestion, it is better to add this assert

File: src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.cpp (hunk 2 above)
Comment (GuoliangShiIntel) [reply src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.cpp]: Done

File: src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.cpp (hunk 3 above)
Comment (GuoliangShiIntel) [reply src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.cpp]: Good proposal, done

File: src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.hpp [hunk 4]
```diff
@@ -76,8 +76,17 @@ class Eagle3Extension {
...
+    // Accumulate last_hidden_state from current chunk during chunked prefill
+    void accumulate_chunk_last_hidden_state(
+        const std::shared_ptr<ov::IAsyncInferRequest>& request,
+        const std::unordered_map<std::string, ov::Output<const ov::Node>>& out_ports,
+        uint32_t chunk_token_count,
+        uint32_t total_seq_len);
+
+    // Reset chunked prefill state before starting a new chunked prefill session
+    void reset_chunked_prefill_state() {
+        m_last_hidden_state = {};
```
Comment (AlexanderKalistratov) [inline src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.hpp:90]: Why we need to do this?
It means on each prefill stage we are allocating a new tensor. Why?

Comment (GuoliangShiIntel) [reply src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.hpp:90]: Good question. Please consider this scenario:

If we run two prompts consecutively using infer:

**For the first prompt:** `m_last_hidden_state` is null -> [pre-allocate tensor for the full tensor](https://github.com/openvinotoolkit/openvino/pull/33975/changes#diff-707a4f44947d8b4903bbbe7a858d2dfea9e04c2dd1be31f9341471927d184acfR186-R187) -> copy each chunk's `last_hidden_state` into pre-allocated memory

After the first prefill completes, the generate phase also [updates](https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.cpp#L150) `m_last_hidden_state`. When the generate phase finishes, `m_last_hidden_state` remains non-null.

**For the second prompt:** Since `m_last_hidden_state` is still non-null, prefill will not enter the "Pre-allocate tensor on first chunk" path, causing a memory size mismatch that [triggers the assertion](https://github.com/openvinotoolkit/openvino/pull/33975/changes#diff-707a4f44947d8b4903bbbe7a858d2dfea9e04c2dd1be31f9341471927d184acfR199-R201).

Given that each prompt inference only prefill once, it's reasonable to reset the tensor here.

Comment (AlexanderKalistratov) [reply src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.hpp:90]: So `m_last_hidden_state` is pointing to different tensors in prefill and generate phase?
It would be nice to have an explanatory comment here.

Having allocation per prefill is not a big deal I think. But we also can have pre-allocated tensor for prefill phase and not allocate it every time.

**Comment by dmatveev** [comment]
build_jenkins

**Comment by GuoliangShiIntel** [comment]
@dmatveev All CI passed, can we merge this PR?

**Comment by dmatveev** [comment]
build_jenkins
